marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -239,227 +268,373 @@ instance FromField C.BlockNo where
-      `union`
-      diskStored
-
--- | Query the data stored in the indexer as a whole from:
-    -- + volatile  : in-memory, datat that may rollback
-    -- + diskStore : on-disk
-    -- + buffered  : in-memeoy, data that will flush to storage
-queryPlusVolatile
-  :: UtxoIndex                  -- ^ in-memory indexer
-  -> C.AddressAny               -- ^ Address to filter for
-  -> IO Result                  -- ^ search results
-queryPlusVolatile ix addr =
-  Ix.getEvents (ix ^. Ix.storage)  >>= query ix addr
-
-onInsert :: UtxoIndex -> UtxoEvent -> IO [Notification]
-onInsert  _ _ =  pure []
-
-store :: UtxoIndex -> IO ()
-store ix = do
-  buffer <- Ix.getBuffer $ ix ^. Ix.storage
-  let rows =  (concatMap toRows) $  buffer
-      spent = concatMap (Set.toList . _utxoEventInputs) buffer
-      conn = ix ^. Ix.handle
-  bracket_
-      (SQL.execute_ conn "BEGIN")
-      (SQL.execute_ conn "COMMIT")
-      ( concurrently_
-            (unless (null rows)
-                (SQL.executeMany conn
-                "INSERT OR REPLACE INTO unspent_transactions (address, txId, txIx, datum, datumHash, value, inlineScript, inlineScriptHash, slotNo, blockNo) VALUES (?,?,?,?,?,?,?,?,?,?)"
-                rows))
-            (unless (null spent)
-                (SQL.executeMany conn
-                "INSERT OR REPLACE INTO spent (txId, txIx) VALUES (?, ?)"
-                spent))
-      )
-  -- We want to perform vacuum about once every 100 * buffer ((k + 1) * 2)
-  rndCheck <- createSystemRandom >>= uniformR (1 :: Int, 100)
-  when (rndCheck == 42) $ do
-    SQL.execute_ conn "DELETE FROM unspent_transactions WHERE unspent_transactions.rowid IN (SELECT unspent_transactions.rowid FROM unspent_transactions LEFT JOIN spent on unspent_transactions.txId = spent.txId AND unspent_transactions.txIx = spent.txIx WHERE spent.txId IS NOT NULL)"
-    SQL.execute_ conn "VACUUM"
-
--- convert to AddressAny from address in any valid cardano era
+instance Queryable UtxoHandle where
+  queryStorage
+    :: Foldable f
+    => QueryInterval C.ChainPoint
+    -> f (StorableEvent UtxoHandle)
+    -> UtxoHandle
+    -> StorableQuery UtxoHandle
+    -> IO (StorableResult UtxoHandle)
+  queryStorage qi es (UtxoHandle c _) q = do
+      persisted <- case qi of
+        QEverything -> SQL.query c
+          [r|SELECT u.address, u.txId, u.txIx, u.datum, u.datumHash
+                  , u.value, u.inLineScript, u.inLineScriptHash
+                  , u.blockNo, u.slotNo, blockHash
+             FROM unspent_transactions u
+             LEFT JOIN spent s ON
+                    u.txId = s.txInTxId
+                AND u.txId IS NOT NULL
+                AND u.txIx = s.txInTxIx
+             WHERE u.address = ?
+             ORDER BY u.slotNo ASC|] (Only q)
+        QInterval _ (C.ChainPoint sn _) -> SQL.query c
+          [r|SELECT u.address, u.txId, u.txIx , u.datum , u.datumHash, u.value
+                  , u.inLineScript, u.inLineScriptHash, u.blockNo, u.slotNo
```

Comment:
Isn't this missing `blockHash`? Do you think we should use a generic query template and add the `slotNo` condition if necessary?